instance/models/rabbitmq_server.py

```diff
@@ -0,0 +1,137 @@
+
+logger = logging.getLogger(__name__)
+
+
+# Constants ###################################################################
+
+DEFAULT_INSTANCE_RABBITMQ_PORT = 5671
+
+
+# Models ######################################################################
+
+class RabbitMQServerManager(models.Manager):
+    """
+    Custom manager for the RabbitMQServer model.
+    """
+
+    def _create_default(self):
+        """
+        Create the default RabbitMQ server configured in the Django settings, if any.
+        """
+        if settings.DEFAULT_RABBITMQ_API_URL:
+            api_url = urlparse(settings.DEFAULT_RABBITMQ_API_URL)
+            amqps_url = urlparse(settings.DEFAULT_INSTANCE_RABBITMQ_URL)
+
+            defaults = {
+                'api_url': api_url.geturl().replace(':'.join([api_url.username, api_url.password]), ''),
+                'admin_username': api_url.username,
+                'admin_password': api_url.password,
+                'instance_host': amqps_url.hostname,
+                'instance_port': amqps_url.port,
+            }
+
+            if not all(defaults.values()):
+                raise ImproperlyConfigured(
+                    "Found DEFAULT_RABBITMQ_API_URL, but it must contain basic auth credentials, and "
+                    "DEFAULT_INSTANCE_RABBITMQ_URL must contain both a hostname and port"
+                )
+            logger.info("Creating RabbitMQServer %s", defaults['instance_host'])
+
+            del defaults['api_url']
+            server, created = self.get_or_create(
+                api_url=settings.DEFAULT_RABBITMQ_API_URL,
+                defaults=defaults,
+            )
+            if not created and not all(getattr(server, name) == default for name, default in defaults.items()):
+                logger.warning(
+                    "RabbitMQServer for %s already exists, and its settings do not match "
+                    "the Django settings: " + ", ".join(["%s vs %s"] * len(defaults)),
+                    defaults['api_url'],
+                    *sum(defaults.items(), ())
+                )
+
+    def select_random(self):
+        """
+        Select a RabbitMQ server for a new instance.
+
+        The current implementation selects one of the RabbitMQ servers that accept new clients at random.
+        If no database server accepts new clients, RabbitMQServer.DoesNotExist is raised.
+        """
+        self._create_default()
+
+        # The set of servers might change between retrieving the server count and retrieving the random server,
+        # so we make this atomic.
+        with transaction.atomic():
+            servers = self.filter(accepts_new_clients=True)
+            count = servers.count()
+            if not count:
+                raise self.model.DoesNotExist(
+                    "No configured RabbitMQServer accepts new clients."
+                )
+            return servers[random.randrange(count)]
```

Comment:
Awesome idea! Especially pulling out the query selector into it's own function. This DID actually come in handy for removing this same duplicated code from the `LoadBalancingServerManager`. :) Thanks a lot!